crates/cranelift/src/func_environ.rs

```diff
@@ -2215,4 +2290,110 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
     fn use_x86_pmaddubsw_for_dot(&self) -> bool {
         self.isa.has_x86_pmaddubsw_lowering()
     }
+
+    cfg_if! {
+        if #[cfg(feature = "valgrind")] {
+            fn handle_before_return(
+                &mut self,
+                retvals: &[Value],
+                builder: &mut FunctionBuilder,
+            ) {
+                if self.valgrind {
+                    let func_index = match &builder.func.name {
+                        UserFuncName::User(user) => {
+                            FuncIndex::from_u32(user.index)
+                        }
+                        _ => {
+                            unreachable!()
+                        }
+                    };
+                    let func_name = self.translation.debuginfo.name_section.func_names[&func_index];
+                    if func_name == "malloc" {
+                        self.check_malloc_exit(builder, retvals);
+                    } else if func_name == "free" {
+                        self.check_free_exit(builder);
+                    }
+                }
+            }
+            
+            fn before_load(&mut self, builder: &mut FunctionBuilder, val_size: u8, addr: ir::Value, offset: u64) {
+                if self.valgrind {
+                    let check_load_sig = self.builtin_function_signatures.check_load(builder.func);
+                    let (vmctx, check_load) = self.translate_load_builtin_function_address(
+                        &mut builder.cursor(),
+                        BuiltinFunctionIndex::check_load(),
+                    );
+                    let num_bytes = builder.ins().iconst(I32, val_size as i64);
+                    let offset_val = builder.ins().iconst(I64, offset as i64);
+                    builder
+                        .ins()
+                        .call_indirect(check_load_sig, check_load, &[vmctx, num_bytes, addr, offset_val]);
+                }
+            }
+
+            fn before_store(&mut self, builder: &mut FunctionBuilder, val_size: u8, addr: ir::Value, offset: u64) {
+                if self.valgrind {
+                    let check_store_sig = self.builtin_function_signatures.check_store(builder.func);
+                    let (vmctx, check_store) = self.translate_load_builtin_function_address(
+                        &mut builder.cursor(),
+                        BuiltinFunctionIndex::check_store(),
+                    );
+                    let num_bytes = builder.ins().iconst(I32, val_size as i64);
+                    let offset_val = builder.ins().iconst(I64, offset as i64);
+                    builder
+                        .ins()
+                        .call_indirect(check_store_sig, check_store, &[vmctx, num_bytes, addr, offset_val]);
+                }
+            }
+
+            fn update_global(&mut self, builder: &mut FunctionBuilder, global_index: u32, value: ir::Value) {
+                if self.valgrind {
+                    if global_index == 0 {
```

Comment:
Let's add a comment here about the assumption we're making -- that global 0 is the auxiliary stack pointer.